pkg/scheduler/ugm/queue_tracker.go

```diff
@@ -403,24 +403,20 @@ func (qt *QueueTracker) canRunApp(hierarchy []string, applicationID string, trac
 // it decides the removal. It returns false the moment it sees any unexpected values for any queue in any levels.
 // Note: Lock free call. The RLock of the linked tracker (UserTracker and GroupTracker) should be held before calling this function.
 func (qt *QueueTracker) canBeRemoved() bool {
-	for _, childQT := range qt.childQueueTrackers {
-		// quick check to avoid further traversal
-		if childQT.canBeRemovedInternal() {
+	if len(qt.childQueueTrackers) > 0 {
+		for _, childQT := range qt.childQueueTrackers {
+			// quick check to avoid further traversal
 			if !childQT.canBeRemoved() {
 				return false
 			}
-		} else {
-			return false
 		}
+		return true // removable because all children are removable
 	}
-	// reached leaf queues, no more to traverse
-	return qt.canBeRemovedInternal()
-}
 
-func (qt *QueueTracker) canBeRemovedInternal() bool {
-	if len(qt.runningApplications) == 0 && resources.IsZero(qt.resourceUsage) && len(qt.childQueueTrackers) == 0 &&
-		qt.maxRunningApps == 0 && resources.IsZero(qt.maxResources) {
-		return true
-	}
-	return false
+	// reached a leaf queue
+	return len(qt.runningApplications) == 0 &&
+		resources.IsZero(qt.resourceUsage) &&
+		len(qt.childQueueTrackers) == 0 &&
+		qt.maxRunningApps == 0 &&
+		resources.IsZero(qt.maxResources)
```

Comment:
@manirajv06 please check this logic, I think this is simpler than the original and accomplishes the same.

Reply:
In newer approach, we take decision only after reaching the leaf queue by traversing all the queues in the queue hierarchy. Whereas earlier, earlier one takes decision at very earlier stage and come out then and there.

Reply:
Even in the old code, `canBeRemovedInternal()` only returns true if `len(qt.childQueueTrackers) == 0`. So you need to descend down the hierarchy. You can't return earlier, therefore the optimization is not really there.

Reply:
For example,

root.parent running app is 10. root.parent has 100's of child queues (or child queue trackers) below it. In earlier approach, We can come out right away the moment we see 10 apps running in root.parent queue and confirm that it is better not to remove root.parent and any queue below it.

Reply:
Ok I'm reverting this change, let's not spend too much time on this. Maybe I'll get back to it later.